integration_test.go

```diff
@@ -506,6 +333,60 @@ func TestErrorHandlingExactlyOnce(t *testing.T) {
 	// Verify the number of rows
 	assert.Equal(t, numGoodRows, rowCount)
 
+	// Clean up - delete the BigQuery dataset and its contents(includes generated table) as well as config & log files
+	cleanUpBQDatasetAndFiles(ctx, t, dataset)
+}
+
+// This function sets up the BigQuery client, and creates a dataset and table in an existing project
+// The configuration file and log file are also created
+func setupBQTableAndFiles(ctx context.Context, t *testing.T, tableSchema bigquery.Schema, exactlyOnceVal string) (*bigquery.Client, *bigquery.Dataset, string, *os.File) {
+	//get projectID from environment
+	projectID := os.Getenv("ProjectID")
+	if projectID == "" {
+		t.Fatal("Environment variable 'ProjectID' is required to run this test, but not set currently")
+	}
+
+	// Set up BigQuery client
+	client, err := bigquery.NewClient(ctx, projectID)
+	if err != nil {
+		t.Fatalf("Failed to create BigQuery client: %v", err)
+	}
+
+	// Create a random dataset and table name
+	datasetHash := randString()
+	datasetID := "testdataset_" + datasetHash
+
+	tableHash := randString()
+	tableID := "testtable_" + tableHash
+
+	// Create BigQuery dataset and table in an existing project
+	dataset := client.Dataset(datasetID)
+	if err := dataset.Create(ctx, &bigquery.DatasetMetadata{Location: "US"}); err != nil {
+		t.Fatalf("Failed to create BigQuery dataset %v", err)
```

Comment:
Yes since t.fatalF uses os.Exit none of the deferred client.Close calls will go through. I added functionality for closing the client before every fatal call (after the client is created).